lib/collection/src/collection/query.rs

```diff
@@ -60,6 +61,83 @@ impl Collection {
         Ok(results.into_iter().next().unwrap())
     }
 
+    /// If the query limit above this value, it will be a subject to undersampling.
+    const SHARD_QUERY_SUBSAMPLING_LIMIT: usize = 128;
+
+    /// Give some more ensurance for undersampling,
+    /// retrieve more points to prevent undersampling errors.
+    /// Errors are still possible, but rare enough to be acceptable compared to
+    /// errors introduced by vector index.
+    const MORE_ENSURANCE_FACTOR: f64 = 1.2;
+
+    /// Creates a copy of requests in case it is possible to apply limit modification
+    /// Returns unchanged requests if limit modification is not applicable.
+    ///
+    /// If there are many independent shards, and we need a very high limit, we can do an optimization.
+    /// Instead of querying all shards with the same limit, we can query each shard with a smaller limit
+    /// and then merge the results. Since shards are independent and data is randomly distributed, we can
+    /// apply probability estimation to make sure we query enough points to get the desired number of results.
+    ///
+    /// Same optimization we already apply on segment level, but here it seems to be even more reliable
+    /// because auto-sharding guarantee random and independent distribution of data.
+    ///
+    /// Unlike segments, however, the cost of re-requesting the data is much higher for shards.
+    /// So we "accept" the risk of not getting enough results.
```

Comment:
If it's so unlikely, can we still do it in case we don't get enough results? 

I imagine a latency spike to be a bit better than getting less results. Getting less results might make the user incorrectly think they have 'reached the end' or 'exhausted all'.

If we basically never expect this to happen, a latency spike might be a 'safer' option.

Wdyt?

Reply:
We can, it is a bit annoying to implement with our batching structure, would be at least twice as many changes